src/tss2-tcti/tcti-cmd.c

```diff
@@ -0,0 +1,453 @@
+    const uint8_t *cmd_buf)
+{
+    TSS2_TCTI_COMMAND_CONTEXT *tcti_cmd = tcti_cmd_context_cast (tcti_ctx);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_cmd_down_cast (tcti_cmd);
+
+    TSS2_RC rc = tcti_common_transmit_checks (tcti_common, cmd_buf, TCTI_COMMAND_MAGIC);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+
+    size_t bytes = fwrite (cmd_buf, 1, size, tcti_cmd->sink);
+    if (bytes != size) {
+    	LOG_ERROR ("Transmitting to subprocess failed: %s",
+    			strerror(errno));
+        return rc;
+    }
+
+    fflush(tcti_cmd->sink);
+
+    tcti_common->state = TCTI_STATE_RECEIVE;
+
+    return rc;
+}
+
+TSS2_RC
+tcti_cmd_cancel (
+    TSS2_TCTI_CONTEXT *tctiContext)
+{
+    TSS2_TCTI_COMMAND_CONTEXT *tcti_cmd = tcti_cmd_context_cast (tctiContext);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_cmd_down_cast (tcti_cmd);
+    TSS2_RC rc;
+
+    rc = tcti_common_cancel_checks (tcti_common, TCTI_COMMAND_MAGIC);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+
+    tcti_common->state = TCTI_STATE_TRANSMIT;
+    tcti_cmd->cancel = 1;
+
+    return rc;
+}
+
+TSS2_RC
+tcti_cmd_set_locality (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    uint8_t locality)
+{
+    TSS2_TCTI_COMMAND_CONTEXT *tcti_cmd = tcti_cmd_context_cast (tctiContext);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_cmd_down_cast (tcti_cmd);
+    TSS2_RC rc;
+
+    rc = tcti_common_set_locality_checks (tcti_common, TCTI_COMMAND_MAGIC);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+
+    tcti_common->locality = locality;
+    return TSS2_RC_SUCCESS;
+}
+// TODO do we need a define for this, this is coupled to device TCTI?
+#define TCTI_ASYNC
+TSS2_RC
+tcti_cmd_get_poll_handles (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    TSS2_TCTI_POLL_HANDLE *handles,
+    size_t *num_handles)
+{
+#ifdef TCTI_ASYNC
+    TSS2_TCTI_COMMAND_CONTEXT *cmd_tcti = tcti_cmd_context_cast (tctiContext);
+
+    if (num_handles == NULL || cmd_tcti == NULL) {
+        return TSS2_TCTI_RC_BAD_REFERENCE;
+    }
+
+    if (handles != NULL && *num_handles < 1) {
+        return TSS2_TCTI_RC_INSUFFICIENT_BUFFER;
+    }
+
+    *num_handles = 1;
+    if (handles != NULL) {
+        handles->fd = fileno(cmd_tcti->source);
+    }
+
+    return TSS2_RC_SUCCESS;
+#else
+    (void)(tctiContext);
+    (void)(handles);
+    (void)(num_handles);
+    return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+#endif
+}
+
+void
+tcti_cmd_finalize(
+    TSS2_TCTI_CONTEXT *tctiContext)
+{
+    TSS2_TCTI_COMMAND_CONTEXT *tcti_cmd = tcti_cmd_context_cast (tctiContext);
+
+    if (tcti_cmd == NULL) {
+        return;
+    }
+
+    int rc = kill(tcti_cmd->child_pid, SIGTERM);
+    if (rc < 0) {
+    	LOG_ERROR("Error shutting down pid(%u): %s",
+    			tcti_cmd->child_pid,
+    			strerror(errno));
+    }
+
+    /*
+     * Probably should do a micro sleep and try waitpid with WNOHANG
+     * if it hasn't exited sent a SIGKILL and reap.
+     * XXX This should be an option...
+     */
+    alarm(5);
```

Comment:
Where is the SIGALRM signal handled?

Reply:
In any syscall below, we don't actually need to explicitly handle it. We send kill and waitpid, indefinitely. This is so waitpid will wake up with EINTR if the process doesn't respond to SIGTERM. If it doesn't respond in 5 seconds, we send SIGKILL and then reap.

I guess their is a race if the alarm comes in before the waitpid gets called, but the solutions get much messier. I don't want to register a handler, or adjust the sigmask, as the main process we're getting linked into, that's usually their thing.
- https://stackoverflow.com/questions/282176/waitpid-equivalent-with-timeout